rust/relay/src/sockets.rs

```diff
@@ -0,0 +1,279 @@
+use anyhow::{bail, Result};
+use std::{
+    collections::HashMap,
+    io,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    task::{ready, Context, Poll},
+    time::Duration,
+};
+use stun_codec::rfc8656::attributes::AddressFamily;
+use tokio::sync::mpsc;
+
+/// A dynamic collection of UDP sockets, listening on all interfaces of a particular IP family.
+///
+/// Internally, [`Sockets`] is powered by [`mio`] and uses a separate thread to poll for readiness of a socket.
+/// Whenever a socket is ready for reading, we send a message to the foreground task which then reads from the socket until it emits [`io::ErrorKind::WouldBlock`].
+pub struct Sockets {
+    /// All currently active sockets.
+    ///
+    /// [`mio`] operates with a concept of [`mio::Token`]s so we need to store our sockets indexed by those tokens.
+    inner: HashMap<mio::Token, mio::net::UdpSocket>,
+
+    /// Which socket we should still be reading from.
+    ///
+    /// [`mio`] sends us a signal when a socket is ready for reading.
+    /// We must read from it until it returns [`io::ErrorKind::WouldBlock`].
+    current_ready_socket: Option<mio::Token>,
+
+    cmd_tx: mpsc::Sender<Command>,
+    event_rx: mpsc::Receiver<Event>,
```

Comment:
```suggestion
    /// Commands that we send to our mio worker thread
    cmd_tx: mpsc::Sender<Command>,
    /// Events that we receive from our mio worker thread
    event_rx: mpsc::Receiver<Event>,
```

And there's no equivalent channel between `Sockets` and its caller because the caller just calls `poll_*`, `bind`, and `unbind`. At first I wasn't sure if one of the channels comes from the `Sockets` caller.